libtiledbsoma/src/soma_reader.cc

```diff
@@ -78,15 +81,33 @@ SOMAReader::SOMAReader(
     std::shared_ptr<Context> ctx,
     std::vector<std::string> column_names,
     std::string_view batch_size,
-    std::string_view result_order)
+    std::string_view result_order,
+    std::optional<std::pair<uint64_t, uint64_t>> timestamp)
     : ctx_(ctx)
-    , uri_(util::rstrip_uri(uri)) {
+    , uri_(util::rstrip_uri(uri))
+    , timestamp_(timestamp) {
     // Validate parameters
     try {
         LOG_DEBUG(fmt::format("[SOMAReader] opening array '{}'", uri_));
         auto array = std::make_shared<Array>(*ctx_, uri_, TILEDB_READ);
+        if (timestamp) {
+            array->set_open_timestamp_start(timestamp->first);
+            array->set_open_timestamp_end(timestamp->second);
+            // FIXME: tiledb::Array::set_open_timestamp_{start,end} are only
+            // effective if called before [re]opening the array. The
+            // tiledb::Array constructor internally opens the array already, so
+            // we need to reopen() for this to have effect.
+            // https://github.com/TileDB-Inc/TileDB/blob/6f92d2864440cec17474f2aae22f223147828c6a/tiledb/sm/cpp_api/array.h#L97
+            // We might need to open the array using the C API and then use the
+            // alternate tiledb::Array constructor that takes ownership of the C
+            // array.
+            array->reopen();
```

Comment:
@gspowley I ran into this weird quirk with the TileDB C++ API, appreciate your input if you have a POV or if I missed something. The way the tiledb::Array constructor already opens the array seems to constrain us to reopen() in order to set the start & end timestamps. (Without using deprecated constructor variants supporting the end timestamp only.)

Reply:
I agree, we need to set the start and end timestamps and reopen the array with the current TileDB C++ API (we have to do this in TileDB-VCF too).